### PR TITLE
Make the command server entirely event based

### PR DESCRIPTION
**Pull request by Keksoj**
Digging into #708 proved difficult because the code of the Command Server lacked a proper refactoring.

This PR renders the event-based nature of the Command Server systematic.

### Replacement of `answer_error()` and `answer_success()`

These functions answered directly to the client. Their use within the code was boiler plate. They are replaced by `return_error()` and `return_success()`, who send a `CommandMessage::Advancement` (new variant) to the command server, using his `command_tx`.

The command server, in turn, calls `notify_advancement_to_client()`, that does what `answer_error()` and `answer_success()` did.

### New enums!

A good-for-all `Success` to be returned either by `Result<Success>` or on a channel. Satisfies `std::fmt::Display` for logging and responding to the client.

A `Response` that wraps `Success` as a `Result` of sorts.

### New structs!

`RequestIdentifier` contains `client_id` and `request_id`, as those go oftentime together.

### Comments

To explain some whats and whys, notably the purpose of the `in_flight` hashmap.

### Documentation

Have a look at the [new doc page](https://github.com/sozu-proxy/sozu/blob/0.14-refactor-command-server/doc/managing_workers.md) about the workflow of the Command Server.

### CODE REVIEW

File: bin/src/command/mod.rs [hunk 1]
```diff
@@ -66,60 +68,86 @@ enum CommandMessage {
...
 }
 
-pub enum OrderSuccess {
-    ClientClose(String),
-    ClientNew(String),
-    ClientRequest,
-    DumpState(CommandResponseData), // contains the cloned state
-    LaunchWorker(u32),
-    ListFrontends(CommandResponseData),
+#[derive(Clone)]
```
Comment (FlorentinDUBOIS) [inline bin/src/command/mod.rs:78]: You could add others traits to it like those:

```
#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
```

File: bin/src/command/mod.rs [hunk 2]
```diff
@@ -66,60 +68,86 @@ enum CommandMessage {
...
+pub struct RequestIdentifier {
+    client: String,  // the client who sent the request (ex: "CL-0")
+    request: String, // the request id (ex: "ID-MAN9QF")
+}
+
+pub enum Response {
```
Comment (FlorentinDUBOIS) [inline bin/src/command/mod.rs:84]: You could add some traits to derive ;)

```
#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
```

File: bin/src/command/mod.rs [hunk 3]
```diff
@@ -66,60 +68,86 @@ enum CommandMessage {
...
+    Error(String),
+    // Todo: refactor the CLI, see issue #740
+    // Processing(String),
+    Ok(Success),
+}
+
+// Indicates success of either inner Sōzu logic and of handling the ClientRequest,
+// in which case Success caries the response data.
+#[derive(Clone)]
```
Comment (FlorentinDUBOIS) [inline bin/src/command/mod.rs:93]: Same comments as above, you could add some traits at least the `Debug`.

File: bin/src/command/mod.rs [hunk 4]
```diff
@@ -589,104 +635,85 @@ impl CommandServer {
...
-            let id = worker.id;
-            let command_tx = self.command_tx.clone();
-            smol::spawn(async move {
-                worker_loop(id, stream, command_tx, worker_rx)
-                    .await
-                    .unwrap();
-            })
-            .detach();
+        let sock = worker.channel.take().unwrap().sock;
+        let (worker_tx, worker_rx) = channel(10000);
```
Comment (FlorentinDUBOIS) [inline bin/src/command/mod.rs:651]: Could you add an underscore between the fifth and the fourth zero like below.

```rust
let (worker_tx, worker_rx) = channel(10_000);
```

File: bin/src/command/mod.rs [hunk 5]
```diff
@@ -1013,17 +998,27 @@ async fn worker_loop(
...
             }
         }
     }
 
-    tx.send(CommandMessage::WorkerClose { id }).await.unwrap();
+    // if the loop breaks, order the command server to close the worker
+    command_tx
+        .send(CommandMessage::WorkerClose { id })
+        .await
+        .unwrap();
```
Comment (FlorentinDUBOIS) [inline bin/src/command/mod.rs:1021]: Could you change it to a `.with_context(...)?`

File: bin/src/command/orders.rs [hunk 6]
```diff
@@ -127,70 +114,56 @@ impl CommandServer {
...
                 counter += 1;
             }
-            f.sync_all()?;
+            file.sync_all()?;
 
             Ok(counter)
         })();
 
-        Ok(res?)
+        Ok(result?)
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:150]: Could you add some context using `.with_context(...)`

File: bin/src/command/orders.rs [hunk 7]
```diff
@@ -127,70 +114,56 @@ impl CommandServer {
...
 
                 if counter % 1000 == 0 {
                     info!("writing command {}", counter);
-                    f.sync_all()?;
+                    file.sync_all()?;
                 }
                 counter += 1;
             }
-            f.sync_all()?;
+            file.sync_all()?;
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:145]: Could you add some context using `.with_context(...)`

File: bin/src/command/orders.rs [hunk 8]
```diff
@@ -127,70 +114,56 @@ impl CommandServer {
...
                         .map(|s| s.into_bytes())
                         .unwrap_or(vec![]),
                 )?;
-                f.write_all(&b"\n\0"[..])?;
+                file.write_all(&b"\n\0"[..])?;
 
                 if counter % 1000 == 0 {
                     info!("writing command {}", counter);
-                    f.sync_all()?;
+                    file.sync_all()?;
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:141]: Could you add some context using `.with_context(...)`

File: bin/src/command/orders.rs [hunk 9]
```diff
@@ -127,70 +114,56 @@ impl CommandServer {
...
                 );
 
-                f.write_all(
+                file.write_all(
                     &serde_json::to_string(&message)
                         .map(|s| s.into_bytes())
                         .unwrap_or(vec![]),
                 )?;
-                f.write_all(&b"\n\0"[..])?;
+                file.write_all(&b"\n\0"[..])?;
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:137]: Could you add some context using `.with_context(...)`

File: bin/src/command/orders.rs [hunk 10]
```diff
@@ -127,70 +114,56 @@ impl CommandServer {
...
                     CommandRequestData::Proxy(command),
                     None,
                 );
 
-                f.write_all(
+                file.write_all(
                     &serde_json::to_string(&message)
                         .map(|s| s.into_bytes())
                         .unwrap_or(vec![]),
                 )?;
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:149]: Could you add some context using `.with_context(...)`

File: bin/src/command/orders.rs [hunk 11]
```diff
@@ -286,21 +259,21 @@ impl CommandServer {
...
-        buffer.available_data(), message_counter, diff_counter);
-        let mut oks_and_errors: [usize; 2] = [0, 0];
         if diff_counter > 0 {
             info!(
                 "state loaded from {}, will start sending {} messages to workers",
                 path, diff_counter
             );
-            let task = smol::spawn(async move {
+
+            let command_tx = self.command_tx.clone();
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:273]: Could you change the `.clone` into `.to_owned`

File: bin/src/command/orders.rs [hunk 12]
```diff
@@ -317,47 +290,55 @@ impl CommandServer {
...
-                                None,
-                            ))
-                            .await
-                        {
-                            error!("could not send message to client {:?}: {:?}", client_id, e);
-                        }
+                // or pattern matching
+
+                let request_identifier = if let Some(client_id) = client_id {
+                    RequestIdentifier {
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:296]: Maybe, we could add a `new` method

File: bin/src/command/orders.rs [hunk 13]
```diff
@@ -757,6 +694,8 @@ impl CommandServer {
                         }
                     };
                 }
+                // shouldn't we use return_success() here, to notify the command server
+                // and the client?
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:698]: It could be a great idea :)

File: bin/src/command/orders.rs [hunk 14]
```diff
@@ -1316,7 +1237,144 @@ impl CommandServer {
...
+
+    pub async fn notify_advancement_to_client(
+        &mut self,
+        request_identifier: RequestIdentifier,
+        response: Response,
+    ) -> anyhow::Result<Success> {
+        let RequestIdentifier {
+            client: client_id,
+            request: request_id,
+        } = request_identifier.clone();
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:1260]: Could you change the method `.clone` to `.to_owned`

File: bin/src/command/orders.rs [hunk 15]
```diff
@@ -1316,7 +1237,144 @@ impl CommandServer {
...
+
+        let command_response: CommandResponse;
```
Comment (FlorentinDUBOIS) [inline bin/src/command/orders.rs:1262]: You should write something more like

```
let command_response = match ... { ... };
```

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]
The documentation 😍

File: bin/src/command/mod.rs (hunk 1 above)
Comment (Keksoj) [reply bin/src/command/mod.rs:78]: Sure, only `Ord` and `PartialOrd` are not implemented by `CommandResponseData`, which is encapsulated but out of the crate. I'll let those out.

File: bin/src/command/mod.rs (hunk 5 above)
Comment (Keksoj) [reply bin/src/command/mod.rs:1021]: Well this is executed in a detached thread, so I'll just log the error in place.

File: bin/src/command/orders.rs (hunk 6 above)
Comment (Keksoj) [reply bin/src/command/orders.rs:150]: sure!

File: bin/src/command/orders.rs (hunk 13 above)
Comment (Keksoj) [reply bin/src/command/orders.rs:698]: Yeah, we will need to solve the issue #740 first. I completed the comments.

**Review by FlorentinDUBOIS — APPROVED** [review]

**Pull request merged**